R/splatPop-simulate.R

```diff
@@ -0,0 +1,849 @@
+            qn.cvs[[group]] <- qn.cv.gr
+        }
+        qn.mean <- rowMeans(as.data.frame(qn.means))
+        qn.cv <- rowMeans(as.data.frame(qn.cvs))
+
+    }else{
+        qn.mean <- rowMeans(MeansMatrix)
+        qn.cv <- apply(MeansMatrix, 1, FUN = co.var)
+    }
+
+    qn.df <- data.frame(list(meanQuantileNorm = qn.mean,
+                             cvQuantileNorm = qn.cv))
+    qn.df$geneID <- row.names(qn.df)
+    key <- merge(key, qn.df, by = "geneID")
+
+    return(key)
+}
+
+
+#' Simulate gene means for splatPop
+#'
+#' Simulate outlier expression factors for splatPop. Genes with an outlier
+#' factor not equal to 1 are replaced with the median mean expression
+#' multiplied by the outlier factor.
+#'
+#' @param sim SingleCellExperiment to add gene means to.
+#' @param params SplatParams object with simulation parameters.
+#' @param base.means.gene List of gene means for sample from matrix
+#'        generated by `splatPopSimulateMeans` and with the sample specified
+#'        in `splatPopSimulateSC`.
+#'
+#' @return SingleCellExperiment with simulated gene means.
+#'
+#' @importFrom SummarizedExperiment rowData rowData<-
+#' @importFrom stats rgamma median
+splatPopSimGeneMeans <- function(sim, params, base.means.gene) {
+
+    # Note: This function is similar to splatSimGeneMeans, except it uses the
+    # simulated gene mean instead of sampling one randomly. If changes are made
+    # to the outlier method for splat, they should also be made here.
+
+    nGenes <- getParam(params, "nGenes")
+    out.prob <- getParam(params, "out.prob")
+    out.facLoc <- getParam(params, "out.facLoc")
+    out.facScale <- getParam(params, "out.facScale")
+
+    # Add expression outliers
+    outlier.facs <- getLNormFactors(nGenes, out.prob, 0, out.facLoc,
+                                    out.facScale)
+    median.means.gene <- median(base.means.gene)
+    outlier.means <- median.means.gene * outlier.facs
+    is.outlier <- outlier.facs != 1
+    means.gene <- base.means.gene
+    means.gene[is.outlier] <- outlier.means[is.outlier]
+
+    rowData(sim)$BaseGeneMean <- base.means.gene
+    rowData(sim)$OutlierFactor <- outlier.facs
+    rowData(sim)$GeneMean <- means.gene
+
+    return(sim)
+}
+
+#' Clean up the population-scale SCE to remove redundant information
+#'
+#' @param sim.all SingleCellExperiment object with counts for all samples
+#'
+#' @return SingleCellExperiment with simulated sc counts.
+#'
+splatPopCleanSCE <- function(sim.all){
+
+    # Remove redundant sce info
+    keep.id <-  gsub("_Gene", "", names(rowData(sim.all))[1])
+
+    shared.out.factor <- rowData(sim.all)[[paste0(keep.id, "_OutlierFactor")]]
+    rowData(sim.all)[grepl("_OutlierFactor", names(rowData(sim.all)))] <- NULL
+    rowData(sim.all)$Shared_OutlierFactor <- shared.out.factor
+
+    rowData(sim.all)[grepl("_Gene", names(rowData(sim.all)))] <- NULL
+    metadata(sim.all)[2:length(names(metadata(sim.all)))] <- NULL
+
+    return(sim.all)
+}
```

Comment:
Can probably replace this with the new `minmiseSCE()` function (which I have written but is in another branch.